common/key_value_database/platform/lmdb/lmdb.cpp

```diff
@@ -0,0 +1,170 @@
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <iostream>
+#include <lmdb++.h>
+#include <cstdio>
+
+#include <common/common.hpp>
+#include <common/key_value_database_lmdb.hpp>
+
+namespace mender::common::key_value_database {
+
+class LmdbTransaction : public Transaction {
+public:
+	LmdbTransaction(lmdb::txn &txn, lmdb::dbi &dbi);
+
+	expected::ExpectedBytes Read(const string &key) override;
+	error::Error Write(const string &key, const vector<uint8_t> &value) override;
+	error::Error Remove(const string &key) override;
+
+private:
+	lmdb::txn &txn_;
+	lmdb::dbi &dbi_;
+};
+
+LmdbTransaction::LmdbTransaction(lmdb::txn &txn, lmdb::dbi &dbi) :
+	txn_(txn),
+	dbi_(dbi) {
+}
+
+expected::ExpectedBytes LmdbTransaction::Read(const string &key) {
+	try {
+		std::string_view value;
+		bool exists = dbi_.get(txn_, key, value);
+		if (!exists) {
+			return MakeError(KeyError, "Key " + key + " not found in database");
+		}
+
+		return common::ByteVectorFromString(value);
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+error::Error LmdbTransaction::Write(const string &key, const vector<uint8_t> &value) {
+	try {
+		string_view data(reinterpret_cast<const char *>(value.data()), value.size());
+		bool exists = dbi_.put(txn_, key, data);
+		if (!exists) {
+			return MakeError(AlreadyExistsError, "Key " + key + " already exists");
+		}
+
+		return error::NoError;
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+error::Error LmdbTransaction::Remove(const string &key) {
+	try {
+		// We don't treat !exists as an error, just ignore return code.
+		dbi_.del(txn_, key);
+
+		return error::NoError;
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+}
+
+KeyValueDatabaseLmdb::KeyValueDatabaseLmdb() :
+	env_(lmdb::env::create()),
+	successfully_opened_(false) {
+}
+
+KeyValueDatabaseLmdb::~KeyValueDatabaseLmdb() {
+	Close();
+}
+
+error::Error KeyValueDatabaseLmdb::Open(const string &path) {
+	Close();
+
+	try {
+		env_.open(path.c_str(), MDB_NOSUBDIR, 0600);
+	} catch (std::runtime_error &e) {
+		return MakeError(LmdbError, e.what());
+	}
+
+	successfully_opened_ = true;
+	return error::NoError;
+}
+
+void KeyValueDatabaseLmdb::Close() {
+	if (!successfully_opened_) {
+		return;
+	}
+
+	env_.close();
+}
+
+expected::ExpectedBytes KeyValueDatabaseLmdb::Read(const string &key) {
+	vector<uint8_t> ret;
+	auto err = ReadTransaction([&key, &ret](Transaction &txn) -> error::Error {
+		auto result = txn.Read(key);
+		if (result) {
+			ret = std::move(result.value());
+			return error::NoError;
+		} else {
+			return result.error();
+		}
+	});
+	if (err) {
```

Comment:
I'm going nuts, cause I was tempted to suggest using `?:` :lab_coat:

Reply:
I thought you would like the Golang style? 🙂

Reply:
haha, I do! 

C++ is ruining me good taste! :scream: :screwdriver: